### PR TITLE
Create Guest Feature

### PR DESCRIPTION
**Pull request by offiongbassey**
Create Guest is to enable Guests save their information. 

guest

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,6 +4,9 @@
   "description": "Guest Management System",
   "main": "index.js",
   "scripts": {
+    "make-guest": "npx sequelize-cli model:generate --name guest --attributes name:string,email:string,phone:string,gender:string",
```
Comment (eleezs) [inline package.json]: Why did you add this script: make-guest?

Use this instead for migrate and down. 
db:migrate": "sequelize db:migrate",
"db:undo:migrate": "sequelize db:migrate:undo",
"db:seed": "sequelize db:seed:all",

its close to the original syntax, that execute.

You don't need the npx, you already using yarn.

File: src/helpers/guest.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+   
+const {name, email, phone, gender} = req.body;
+try {
+const newGuest = await guest.create({
+    name, 
+    email,
+    phone,
+    gender
+});
+res.status(201).json({success: true, statusCode: 201, message: "Guest Information successfully saved", Guest: newGuest});
```
Comment (eleezs) [inline src/helpers/guest.js]: write a function to handle all your response.

File: src/server/config/config.js [hunk 3]
```diff
@@ -1,26 +1,26 @@
 require('dotenv').config();
 
-const {database, username, password, host} = process.env;
+const {DB_USER, DB_NAME, DB_PASSWORD, DB_HOST} = process.env;
```
Comment (eleezs) [inline src/server/config/config.js]: {DB_USER, DB_NAME, DB_PASSWORD, DB_HOST} 

format properly { DB_USER, DB_NAME, DB_PASSWORD, DB_HOST }

**Review by eleezs — CHANGES_REQUESTED** [review]
I have left comments

File: src/helpers/guest.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import {guest} from "../server/models";
+import asyncHanlder from "express-async-handler";
```
Comment (eleezs) [inline src/helpers/guest.js]: why this?

File: src/index.js [hunk 5]
```diff
@@ -2,18 +2,18 @@ import express from "express";
...
+import errorHandler from "../src/middlewares/errorHandler";
 dotenv.config();
 
 const app = express();
-app.use(express.urlencoded({extended: true}));
-
+app.use(express.json());
 app.use(cors());
-
+app.use('/api/guest', guestRoute);
```
Comment (eleezs) [inline src/index.js]: Let me suggest a better format

app.use('/api/', router);

Let this reference your index file in your route folder.

File: src/middlewares/validationHandler/guestValidator.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { check, validationResult } from "express-validator";
+
+exports.guestValidator = [
```
Comment (eleezs) [inline src/middlewares/validationHandler/guestValidator.js]: Format properly please

Comment (eleezs) [inline src/middlewares/validationHandler/guestValidator.js]: The variable should be in snake case, this is an array not a function. Do you get?

File: src/middlewares/validationHandler/guestValidator.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    check('name').trim().not().isEmpty()
+    .withMessage("Please provide your Name")
+    .isLength({min: 5, max: 50})
+    .withMessage("Name must be within 5 to 50 Characters"),
+    check('email').normalizeEmail().isEmail()
+    .withMessage("Please provide a valid Email"),
+    check('phone').trim().not().isEmpty()
+    .withMessage("Please provide your Phone Number")
```
Comment (eleezs) [inline src/middlewares/validationHandler/guestValidator.js]: What if the pass +234, or any country code

Comment (eleezs) [inline src/middlewares/validationHandler/guestValidator.js]: Are you checking if the phone number and email already exist, to prevent duplicate records

File: src/middlewares/validationHandler/guestValidator.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    .isLength({min: 11, max: 11})
+    .withMessage("Please provide a valid Phone Number"),
+    check('gender').not().isEmpty()
+    .withMessage("Please Select your Gender")
+    
+]
+
+exports.guestValidation = (req, res, next) => {
```
Comment (eleezs) [inline src/middlewares/validationHandler/guestValidator.js]: Move this to a helper file, all custom functions from your validators should be in a helper file, for easy readability, and reference

File: src/routes/guest-route.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import express from "express";
+const router = express.Router()
+import {createGuest} from "../helpers/guest";
+import {guestValidator, guestValidation} from "../middlewares/validationHandler/guestValidator";
+
+router.post('/', guestValidator, guestValidation, createGuest);
```
Comment (eleezs) [inline src/routes/guest-route.js]: this route is not great. You can do better.

File: src/middlewares/validationHandler/guestValidator.js (hunk 8 above)
Comment (eleezs) [inline src/middlewares/validationHandler/guestValidator.js]: Let this function take an argument, i.e the array you validating.
So in the route all you have to do is call this function and pass the array into for it to validate it.

File: package.json (hunk 1 above)
Comment (offiongbassey) [reply package.json]: Okay. I have resolved that.

File: src/helpers/guest.js (hunk 2 above)
Comment (offiongbassey) [reply src/helpers/guest.js]: I have written a function that handles all the responses.

File: src/server/config/config.js (hunk 3 above)
Comment (offiongbassey) [reply src/server/config/config.js]: Done

File: src/helpers/guest.js (hunk 4 above)
Comment (offiongbassey) [reply src/helpers/guest.js]: asyncHandler handles the asynchronous operations. But I have removed it. Async alone will handle it.

File: src/index.js (hunk 5 above)
Comment (offiongbassey) [reply src/index.js]: Okay. I have done that as well.

File: src/middlewares/validationHandler/guestValidator.js (hunk 6 above)
Comment (offiongbassey) [reply src/middlewares/validationHandler/guestValidator.js]: Yes I have modified it.

File: src/middlewares/validationHandler/guestValidator.js (hunk 7 above)
Comment (offiongbassey) [reply src/middlewares/validationHandler/guestValidator.js]: I felt it wasn't necessary. However, I have resolved it.

File: src/middlewares/validationHandler/guestValidator.js (hunk 8 above)
Comment (offiongbassey) [reply src/middlewares/validationHandler/guestValidator.js]: I have moved it.

File: src/middlewares/validationHandler/guestValidator.js (hunk 7 above)
Comment (offiongbassey) [reply src/middlewares/validationHandler/guestValidator.js]: I have validated Phone Number and Email to prevent duplicate records.

File: src/middlewares/validationHandler/guestValidator.js (hunk 8 above)
Comment (offiongbassey) [reply src/middlewares/validationHandler/guestValidator.js]: Okay. I have done that.

File: src/routes/guest-route.js (hunk 9 above)
Comment (offiongbassey) [reply src/routes/guest-route.js]: I have modified it. Thanks.

File: src/middlewares/validationHandler/guestValidator.js (hunk 6 above)
Comment (offiongbassey) [reply src/middlewares/validationHandler/guestValidator.js]: Okay. I have.

File: src/helpers/guestValidator.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import { body } from "express-validator";
+import {existingEmail, existingPhone, checkAllowedFields} from "../helpers/validation";
+
+export const guest_validator = [
+    body('name')
+        .notEmpty()
```
Comment (eleezs) [inline src/helpers/guestValidator.js:8]: first, make sure this parameter exists. 
.exists()
.withMessage('')

Always check if it exists.

File: src/server/models/guest.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+const {
+  Model
+} = require('sequelize');
+module.exports = (sequelize, DataTypes) => {
+  class Guest extends Model {
```
Comment (eleezs) [inline src/server/models/guest.js:6]: If your migration has the table name Guests
Why is your table name on model then: Guest

File: src/server/migrations/20230723192259-create-guest.js
Comment (eleezs) [inline src/server/migrations/20230723192259-create-guest.js:1]: WHy do you have 2 guest migrations?

File: src/controllers/guest.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import {responseHandler} from "../helpers/responseHandler";
+import {errorHandler} from "../helpers/errorHandler";
+
+export const createGuest = async(req, res) => {
+    try {
+        const newGuest = await Model.Guest.create(req.body);
+        responseHandler(res, 201, true, "Guest Information successfully saved", newGuest);
+    } catch (error) {
+        responseHandler(res, 500, false, "Something went wrong, please try again later")
+        return errorHandler(error);
```
Comment (eleezs) [inline src/controllers/guest.js]: if it the errorHandler is async, await else call it first before returning the a response
await errorhandler(err)
return responseHandler(res, 500, false, "Something went wrong, please try again later")

File: src/helpers/validation.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import Model from "../server/models";
+export const existingEmail = async (email, { req }) => {
```
Comment (eleezs) [inline src/helpers/validation.js:3]: I need space here

File: src/helpers/validation.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  const check_email_existence = await Model.Guest.findOne({ where: { email } });
+  if (check_email_existence) {
+    throw new Error("Sorry, email already exist");
+  }
+  return true;
+};
+
+export const existingPhone = async (phone, { req }) => {
```
Comment (eleezs) [inline src/helpers/validation.js]: Why are you passing req here when you not using it?

**Review by eleezs — CHANGES_REQUESTED** [review]

File: src/helpers/validation.js (hunk 14 above)
Comment (offiongbassey) [reply src/helpers/validation.js]: I have removed it.

File: src/helpers/validation.js (hunk 13 above)
Comment (offiongbassey) [reply src/helpers/validation.js:3]: Done

File: src/controllers/guest.js (hunk 12 above)
Comment (offiongbassey) [reply src/controllers/guest.js]: I have resolved it.

File: src/server/migrations/20230723192259-create-guest.js
Comment (offiongbassey) [reply src/server/migrations/20230723192259-create-guest.js:1]: That was a mistake. I have deleted one.

File: src/server/models/guest.js (hunk 11 above)
Comment (offiongbassey) [reply src/server/models/guest.js:6]: I have resolved it.

File: src/helpers/guestValidator.js (hunk 10 above)
Comment (offiongbassey) [reply src/helpers/guestValidator.js:8]: Okay. I have done that.

File: src/server/migrations/20230726233132-create-guest.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
+'use strict';
```
Comment (eleezs) [inline src/server/migrations/20230726233132-create-guest.js:1]: your table name should be plural

Comment (offiongbassey) [reply src/server/migrations/20230726233132-create-guest.js:1]: Okay I have done that.
